src/runtime/reactor.rs

```diff
@@ -128,8 +128,58 @@ impl Reactor {
 
     /// Block until at least one pending pollable is ready, waking a pending future.
     pub(crate) fn block_on_pollables(&self) {
+        self.check_pollables(|targets| {
+            debug_assert_ne!(
+                targets.len(),
+                0,
+                "Attempting to block on an empty list of pollables - without any pending work, no progress can be made and wasi::io::poll::poll will trap"
+            );
+            wasi::io::poll::poll(targets)
+
+        })
+    }
+
+    /// Without blocking, check for any ready pollables and wake the
+    /// associated futures.
+    pub(crate) fn nonblock_check_pollables(&self) {
+        // Lazily create a pollable which always resolves to ready.
+        use std::sync::LazyLock;
+        static READY_POLLABLE: LazyLock<Pollable> =
+            LazyLock::new(|| wasi::clocks::monotonic_clock::subscribe_duration(0));
+
+        self.check_pollables(|targets| {
+            // Create a new set of targets, with the addition of the ready
+            // pollable:
+            let ready_index = targets.len();
+            let mut new_targets = Vec::with_capacity(ready_index + 1);
+            new_targets.extend_from_slice(targets);
+            new_targets.push(&*READY_POLLABLE);
+
+            // Poll is now guaranteed to return immediately, because at least
+            // one member is ready:
+            let mut ready_list = wasi::io::poll::poll(&new_targets);
+
+            // Erase our extra ready pollable from the ready list:
+            ready_list.retain(|e| *e != ready_index as u32);
+            ready_list
+        })
+    }
+
+    /// Common core of blocking and nonblocking pollable checks. Wakes any
+    /// futures which are pending on the pollables, according to the result of
+    /// the check_ready function.
+    fn check_pollables<F>(&self, check_ready: F)
+    where
+        F: FnOnce(&[&Pollable]) -> Vec<u32>,
+    {
         let reactor = self.inner.borrow();
 
+        // If no wakers are pending on pollables, there is no work to be done
+        // here:
+        if reactor.wakers.is_empty() {
+            return;
+        }
```

Comment:
This one caught me off guard a little. I assume it's correct, but I'm not sure *why* it's correct. Can you elaborate a little more in the comment?

On line 135 we panic if the list of pollables is empty, but apparently having an empty list of wakers is not a problem. Saying a little more about that would be helpful to me at least.

Reply:
This is a good catch, this would have subtly introduced bad behavior in the blocking case, so I added a commit that changes how this is implemented. In the nonblocking case (i.e. the root future is pending but has been woken), we want to skip these calculations and poll call if it cannot wake any other futures. In the blocking case, we want to panic with the most informative message possible, rather than rely on a debug assert, or let wasi trap and the user have to work out why.